services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/CopyRequest.java

```diff
@@ -0,0 +1,227 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import static software.amazon.awssdk.utils.Validate.paramNotNull;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import software.amazon.awssdk.annotations.NotThreadSafe;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * Copy the given {@link AsyncRequestBody} to an object in S3. For file-based copys, you may use {@link CopyFileRequest} instead.
+ *
+ * @see S3TransferManager#copy(CopyRequest)
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class CopyRequest
+    implements TransferObjectRequest,
+               ToCopyableBuilder<CopyRequest.Builder, CopyRequest> {
+
+    private final CopyObjectRequest copyObjectRequest;
+    private final TransferRequestOverrideConfiguration configuration;
+
+    private CopyRequest(DefaultBuilder builder) {
+        this.copyObjectRequest = paramNotNull(builder.copyObjectRequest, "copyRequest");
+        this.configuration = builder.configuration;
+    }
+
+    /**
+     * @return The {@link CopyObjectRequest} request that should be used for the copy
+     */
+    public CopyObjectRequest copyObjectRequest() {
+        return copyObjectRequest;
+    }
+
+    /**
+     * @return the optional override configuration
+     * @see Builder#overrideConfiguration(TransferRequestOverrideConfiguration)
+     */
+    @Override
+    public Optional<TransferRequestOverrideConfiguration> overrideConfiguration() {
+        return Optional.ofNullable(configuration);
+    }
+
+    /**
+     * Create a builder that can be used to create a {@link CopyRequest}.
+     *
+     * @see S3TransferManager#copy(CopyRequest)
+     */
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+    public static Class<? extends Builder> serializableBuilderClass() {
+        return DefaultBuilder.class;
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return new DefaultBuilder(this);
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
Can we add tests for equals and hashcode for all model classes? We've run into issues before where we forgot to update them after we added a new field